### PR TITLE
Support subdirectories in articles

### LINKED ISSUE #205: Links do not work on StackExchange archives
It seems to be because the articles are not all at the same directory level

### PR DESCRIPTION
**Pull request by mossroy**
Please do not merge for now.

I create the PR to have some space to discuss.
I've implemented the idea of the "base" tag, that should fix #205 
But it's still quick and dirty

### CODE REVIEW

**Comment by Jaifroid** [comment]
Thank you for this, @mossroy , it looks like a more complete and generic solution. I'll do some testing.

File: www/js/app.js [hunk 1]
```diff
@@ -813,8 +825,16 @@ define(['jquery', 'zimArchiveLoader', 'util', 'uiUtil', 'cookies','abstractFiles
...
+                // Compute current link's url (with its namespace), if applicable
+                var fullUrl = this.href;
+                var urlMatch = fullUrl.match(regexpUrl);
+                var url;
+                if (urlMatch) {
+                    url = urlMatch[0];
+                }
+                else {
+                    url = $(this).attr("href");
+                }
```
Comment (Jaifroid) [inline www/js/app.js]: @mossroy I couldn't resist trying all this as a one-liner....!
`var url = regexpUrl.test(this.href) ? this.href.match(regexpUrl)[0] : $(this).attr("href");`
I'm half joking, but perhaps there's no need to define `fullUrl`, when you could just match against `this.href`.

File: www/js/lib/zimArchive.js [hunk 2]
```diff
@@ -256,10 +254,6 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
      */
     ZIMArchive.prototype.getDirEntryByTitle = function(title) {
         var that = this;
-        // If no namespace is mentioned, it's an article, and we have to add it
-        if (regexpTitleWithoutNameSpace.test(title)) {
-            title= "A/" + title;
-        }
```
Comment (Jaifroid) [inline www/js/lib/zimArchive.js]: Nice that you could clean this up!

**Comment by Jaifroid** [comment]
@Mossroy PS I forgot to say I've tested this branch/PR against a couple of Stackexchange ZIMs, and a couple of Wikipedia ZIMs, and it all seems to function as intended. Nice work!

(I haven't tested ServiceWorker mode.)

**Comment by Jaifroid** [comment]
@mossroy This just occurred to me:
The regex
`var regexpUrl = /(?:[-ABIJMUVWX]\/.*)$/; `
matches this string:
`Synthesis of RNA/12345.html`
And the matched part is:
`A/12345.html`
So we need a way to avoid that kind of rare but possible match of a capital letter at the end of part of the string. Your regex also matches `A/` (the match is `A/`), which might not be a problem but could be avoided by using `.+` instead of `.*`, to ensure at least one character after the forward slash.

Finally, I think the end-of-string anchor is redundant.

Consider rewriting as:

`var regexpUrl = /(?:^|\/)([-ABIJMUVWX]\/.+)/;`

The match, if any, will then be stored in match[1] rather than match[0]. By putting `(?:^|\/)` at the beginning, the regex will match either the start of the string followed by the pattern we want OR a / anywhere in the string followed immediately by the pattern we want. That way we don't match `Synthesis of RNA/12345.html` either, but we would match `Synthesis of RN/A/12345.html`. I tested this quickly and it seems to work (if you try, remember to change `urlMatch[0]` to `urlMatch[1]`).

**Comment by mossroy** [comment]
Thanks for all this valuable feedback.
I agree with you and will update this PR (probably this weekend)

**Comment by mossroy** [comment]
I pushed a commit with the changes you suggested. But it's not ready to merge yet.
The ServiceWorker mode seems to be broken, with no big surprise.
I'll try to have a look tomorrow.

**Comment by mossroy** [comment]
I've pushed another commit that fixes some issues in ServiceWorker mode, and does a bit more code cleanup.
There are still bugs in ServiceWorker mode when subdirectories are used for articles. But it's not worse than before. I've created #312 to solve that later

**Comment by mossroy** [comment]
I won't have more time to work on that today.
But it's probably worth merging it as it solves the issue in jQuery mode.
@Jaifroid : if it's good for you, I might rebase/squash.

I'll probably have a bit of time to work on kiwix-js by the end of the first week of January : maybe it would be worth releasing a version 2.2 at that time, that would include all the latest improvements we commited? That would in particular allow to use stackexchange ZIM files in browser extensions and Ubuntu touch.

**Comment by Jaifroid** [comment]
@mossroy Would be good to work towards a 2.2 at end Jan. 👍

**Pull request merged**